packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -0,0 +1,386 @@
+    address issuer,
+    address account,
+    uint256 issuedOn,
+    address signer,
+    uint8 v,
+    bytes32 r,
+    bytes32 s
+  ) public isValidUser(issuer, account) {
+    require(
+      isValidAttestation(identifier, issuer, account, issuedOn, signer, v, r, s),
+      "Signature is invalid"
+    );
+    for (uint256 i = 0; i < identifierToAddresses[identifier][issuer].length; i = i.add(1)) {
+      // This enforces only one attestation to be uploaded
+      // for a given set of (identifier, issuer, account)
+      // Editing/upgrading an attestation requires that it be deleted before a new one is registered
+      require(
+        identifierToAddresses[identifier][issuer][i].account != account,
+        "Attestation for this account already exists"
+      );
+    }
+    OwnershipAttestation memory attestation = OwnershipAttestation(account, issuedOn, signer);
+    identifierToAddresses[identifier][issuer].push(attestation);
+    addressToIdentifiers[account][issuer].push(identifier);
+    emit AttestationRegistered(identifier, issuer, account, issuedOn, signer);
+  }
+
+  /**
+   * @notice Deletes an attestation 
+   * @param identifier Hash of the identifier to be deleted
+   * @param issuer Address of the attestation issuer
+   * @param account Address of the account mapped to the identifier
+   * @dev Throws if sender is not the issuer, account, or an authorized AttestationSigner
+   */
+  function deleteAttestation(bytes32 identifier, address issuer, address account)
+    public
+    isValidUser(issuer, account)
+  {
+    OwnershipAttestation[] memory attestations = identifierToAddresses[identifier][issuer];
+    for (uint256 i = 0; i < attestations.length; i = i.add(1)) {
+      OwnershipAttestation memory attestation = attestations[i];
+      if (attestation.account == account) {
+        // This is meant to delete the attestation in the array
+        // and then move the last element in the array to that empty spot,
+        // to avoid having empty elements in the array
+        // TODO reviewers: is there a better way of doing this?
+        identifierToAddresses[identifier][issuer][i] = attestations[attestations.length - 1];
+        identifierToAddresses[identifier][issuer].pop();
+
+        bool deletedIdentifier = false;
+
+        bytes32[] memory identifiers = addressToIdentifiers[account][issuer];
+        for (uint256 j = 0; j < identifiers.length; j = j.add(1)) {
+          if (identifiers[j] == identifier) {
+            addressToIdentifiers[account][issuer][j] = identifiers[identifiers.length - 1];
+            addressToIdentifiers[account][issuer].pop();
+            deletedIdentifier = true;
+            break;
+          }
+        }
+        // Should never be false - both mappings should always be updated in unison
+        assert(deletedIdentifier);
+
+        emit AttestationDeleted(identifier, issuer, account);
+        break;
+      }
+    }
+  }
+
+  function revokeSigner(address signer) public {
```

Comment:
We should probably have some function for cleaning up the attestations that are revoked. It could look very similar to `deleteAttestation`, but would not require the caller to have any per    missions. It may be the case that no one will call this function, but it if it is gas positive via refunds for cleared storage, then folks might (or an issuer can call it themselves to clean things up). In addition to a standalone cleanup function, we may add deleting     revoked attestations as a step in the register or delete functions. This would be an easy choice if it turns out to be gas positive.

Reply:
Yeah I think that's reasonable, could do this either by:
1. modifying the `deleteAttestation` function directly and checking permissions only after the attestation has been found (`require(msg.sender == ... || isRevoked(attestation.signer))`
2. including batch deletion functions, described a bit [here](https://www.notion.so/clabsco/ASv2-On-chain-Registry-Protocol-47e7850656444c37973856e0634ee2f9#505e2c06a162428c85f4c6ff142eb581) to delete all revoked attestations per (`identifier`, `issuer`) or (`account`, `issuer`)

I think (2) would be convenient but may end up not offering a large refund if the caller ends up iterating through mostly non-revoked attestations? (1) could be used after looking up all attestations, check revocation status once per returned signer, revoke attestations using (1) for attestations that have been revoked

wdyt on this?

Reply:
I think those are good approaches. Because `deleteAttestaion` currently iterates through the array of attestations, it runs in expected linear time with the number attestations for a given `(identifier, issuer)`, keeping in mind that the most common number is _probably_ 1. Barring a change suggested below, I'd say 2 is preferable because you are essentially going to iterate through the array anyway.

What I did think of while writing this is that it would be quite useful to be able to pass in an index to the `deleteAttestation` function so that the function doesn't need to iterate through the array at all. You would essentially run the view call off-chain to identify the index of the attestation you want to delete, then send the index along as an (optional) argument and the intended attestations will be accessed directly.

Reply:
Documentation around relevant decisions here:
- we moved from deletion + signer-based revocation -> attestation-based revocation (i.e. no separation between deletion/revocation) + implemented batchRevocation to help in the case of a compromised signer
  - see [PR description for batchRevocation](https://github.com/celo-org/celo-monorepo/pull/9647)
- we added limits to the array sizes (for attestations and identifiers) to be able to make certain guarantees about batchRevocation; see [this PR for more details + more thorough gas benchmarking of batchRevocation](https://github.com/celo-org/celo-monorepo/pull/9656)